Apply desktop Customisations to all users

C0D3-M4513R opened the pull request:
**Status**: Testing
- [x] LXDE
- [x] MATE
- [x] LXQt
- [x] Xfce
- [x] GNUstep

Also please link the issues below.
**Reference**: fixes #4014 fixes #3994

C0D3-M4513R committed:
Apply LXDE custumisations to all users

C0D3-M4513R committed:
Add Desktop shortcuts to all users

C0D3-M4513R committed:
Merge branch 'dev' into patch-1

C0D3-M4513R committed:
Apply default configs for LXQt and MATE to all users

C0D3-M4513R committed:
Stuff was pulled from wrong repo

A couple instances in dietpi-software contained github links to MichaIng/Dietpi, instead of the config specified

MichaIng commented:
Ah, as said, I aim to remove the `dietpi` user, while keeping only the group as pure system group to share cross-software access. We'd need to find a way to either override user defaults system-wide or otherwise copy/link those files and configs when a certain user is chosen for desktop auto-login via `dietpi-autostart`.

C0D3-M4513R commented:
how about creating a folder with all the default user configs, that get copied, once a autostart-user is selected?
eg. `/usr/share/dietpi/defaultuserconfig/` which gets copied to `$USER_HOME/.config/`?
EDIT: I'll implement this later, since you suggested it in the issue already.

MichaIng commented:
For the configuration files there is probably a local user-wide override directory already, so that nothing needs to be copied. Not sure about the desktop entries. What needs to be considered is that most DietPi scripts require root permissions, so executing them from desktop as non-root fails. That would require own entries with a sudo prefix, or we could skip them to require a root desktop or shell session.

C0D3-M4513R committed:
Save desktop configs to default conf dir

C0D3-M4513R committed:
Add desktop programms only on root desktop

C0D3-M4513R committed:
Automatically add pcmanfm File Manager to Desktop

C0D3-M4513R committed:
Prevent linkting the FileManager twice

C0D3-M4513R commented:
Acidentially deleted the branch. I should really start naming my pr branches!

MichaIng commented on dietpi/dietpi-autostart:
```diff
@@ -170,7 +170,9 @@ NB: You might need to grant the user permissions to run the chosen autostart opt
-
+			elif (($ID_AUTOSTART == 2)); then
+				# File manager desktop icon
+				[[ -f '/usr/share/applications/pcmanfm.desktop' ]] && G_EXEC ln -sf /usr/share/applications/pcmanfm.desktop /root/Desktop/pcmanfm.desktop
```
What was the idea here? Adding PCManFM desktop entry should still only be done within LXDE install step since other desktops have other file managers. Also one would want to have it regardless of autologin or LightDM login, manual `startx` call or VNC/XRDP, right?

What we could do, is similarly to how it's done in case of CAVA, looping through known desktop-related desktop entries `/root/Desktop/*.desktop` (only the ones which we link within any shared or individual desktop install step) and link those to the chosen users desktop dir, if there is no existing entry yet. So when one alters or removes entries from the root desktop, these changes are ported to new users as well, and if something is changed on a new users desktop, it is not overwritten, only missing entries are re-created.

C0D3-M4513R replied:
What if the user is actually using the root user, as a means to configure the system? Then that system would break down. I didn't do anything, partly because I was busy, but I also do not like that idea. How about just letting default applications be at `/usr/share/applications/`, and then doing, what you suggested?

Also the `pcfmanfm.desktop` icon will only be copied, if the file-manager is actually installed. 
~~What is done here, should also be done for this.~~
No, everything in the applications folder should be linked/copied? once!
https://github.com/MichaIng/DietPi/blob/d44e9aad4b3b1f10ef1adc0be14054ac4b1b9c14/dietpi/dietpi-software#L9157 -

Also, yes, this should be on all desktop-logins, but I dunno, how to handle a login-manager (lightdm, gdm3 etc.), since a login-manager could login any user, and I wouldn't be able to check, which user logged in, and if they already have the desktop icons in place.

Also, yes, the links should only be created once, but how do I know, that they have been copied once? (`touch` a file, to check after?)
Furthermore, how do I detect changes to `/usr/share/applications` (which shouldn't happen?) ?

This could both be solved using a default user systemd service, dedicated to copying the config-files once, and then creating a file, which disables the service via a start-conditional (pre-start check, if the service should start).
I will have to check, where to place the systemd-unit, for it to be a default-user unit, but otherwise this is doable.

MichaIng replied:
True, not easy via `dietpi-autostart` to get that right. A different very simple approach:
```
/etc/skel/Desktop/htop.desktop
```
The skeleton directory is copied automatically in any new users home directory. This includes system users as long as via adduser it is not explicitly unset or via useradd explicitly set. But the desktop entries are symlinks and the only other left file `.gtk-bookmarks` is tiny, so no large overhead, if a system user gets those additional files into his home.
On software installs, we can then additionally loop through user homes and add the one desktop icon that is to be installed. dietpi-software/config/launcher btw only makes sense on root's home since those require root permissions anyway. That way it is guaranteed that every file is only created once and every now and existing user get's once the entries which we think are helpful enough. What do you think?
One issue is that on reinstall, desktop icons for the reinstalled software title are re-created, as long as we do not place a flag file somewhere.
_____
Not sure if I understood your idea with the systemd service right. A systemd unit can disable itself e.g. via ExecStartPost, so that it will start only once or when manually re-enabled/started. They can also be invoked as `--user` services, although in this case it could be a regular root system service that applies correct permissions after copying files. But how do you know which one is the default user that needs to have files copied?

C0D3-M4513R replied:
Ok, so going upon that, we could create a global user service in `/etc/systemd/system` , which starts on login, if there is no `~/.gtk-bookmarks`, and then the service copies all the files (as the action of the unit).
This would solve the problem, of copying files, where they shouldn't be, since systemd afaik only starts on users, which can log in (so actual human users, and not service users).

MichaIng replied:
I'm still not sure how to implement, since there is no systemd target for an actual user login 🤔. The systemd boot sequence ends with the TTY1 (what you see on the locally attached monitor, if there is one) login prompt, i.e. the `getty@tty1.service`. What happens after that, cannot be automated via systemd services, from all I know. But there are shell and desktop startup script/applications:
- `/etc/profile`, `/etc/profile.d/*.sh` and `~/.profile` are loaded whenever a user logs into a login shell, e.g. via local console or SSH.
- `/etc/bash.bashrc`, `/etc/bashrc.d/*.sh|.bash` and `~/.bashrc` are loaded whenever a user opens a bash shell, so including a non-login shell, subshell or terminal emulator from within a desktop session.
- But what fits best here: `/etc/xdg/autostart/*.desktop` and `~/.config/autostart/*.desktop`: https://specifications.freedesktop.org/autostart-spec/autostart-spec-latest.html
    A desktop autostart entry indeed starts up when the desktop environment is loaded. And a systemd wide entry can be disabled by creating a same named file in the user autostart dir which contains `Hidden=true`.
    So in theory we could make that desktop entry creation script a system-wide autostart entry which automatically disables itself on user-level, once files have been copied. Indeed that would rule out files being created for system users or even regular users which are just not used to log into desktop environments. That way we can indeed skip pre-creating any desktop entries or configs for any user, including root. And we don't need to check for a flag file or such. The autostart script just needs to loop through our "known" desktop entries/configs, and link/copy them, if they exist on system-level and do not exist already on user-level (should not happen, but theoretically a user could manually pre-create it's desktop icons, e.g. from a home dir backup or such). For `.gtk-bookmarks`, we'd need to download and store it somewhere on the system, e.g. `/var/lib/dietpi/dietpi-software/desktops/.gtk-bookmarks` at install.

C0D3-M4513R replied:
I think `user.slice` is created, when a user logs in.
Also, if a service is a user service, it won't get run(or even be tried to run), until a user actually loggs in.
You have the option of deciding, weather we do a `.desktop` autostart, or a systemd service.
Personal preference is systemd autostart, since that feels cleaner.

MichaIng replied:
Good find and good to know! But it is active as well on console logins, so that desktop autostart entry would allow to easily apply things on actual desktop sessions only. I'll implement this.

C0D3-M4513R replied:
You can set services to start after `graphical.target` (which would be desktop) , or  `multiuser.target` (which would be console).
Also should we close this pr, or do you just want to take over from here (In this option, you should assign yourself to this pr)?

MichaIng replied:
`graphical.target` is also reached when no one logs into a desktop, even when the system is completely headless:
```
2021-02-02 12:12:08 root@micha:/tmp# systemctl status graphical.target
● graphical.target - Graphical Interface
     Loaded: loaded (/lib/systemd/system/graphical.target; static)
     Active: active since Mon 2021-02-01 21:55:01 CET; 14h ago
       Docs: man:systemd.special(7)

Feb 01 21:55:01 my.domain.org systemd[1]: Reached target Graphical Interface.
```
What gives this target a purpose is the `display-manager.service`, if it is present and enabled. This can be theoretically a desktop autostart, when manually setup that way, but is usually the LightDM (or other diplay/login manager) login prompt, but not the login itself. So using that target, the way I currently think of, would fire our script at every boot, either when LightDM prompt appears or, if not enabled, when the console login appears, without giving us a user context yet. In the latter case `graphical.target` and `multiuser.target` are reached directly one after another, without any meaning:
```
Jan 30 00:31:45 my.domain.org systemd[1]: Reached target Multi-User System.
Jan 30 00:31:45 my.domain.org systemd[1]: Reached target Graphical Interface.
```
But maybe I'm too bound in the way I think of systemd, and to be true I have not much experiences with user-level services and slices. So if you want to play with it and come up with a PR (or even this one), I would be happy to migrate things to a systemd-based solution, which would allow likely more fine grained features and ties into the general boot and login order, which is all based on systemd. But the only concrete solution I can directly implement is the desktop entry autostart one, even that it is not that nicely hidden from the UI and feels a bit like a misuse/cheat 😄.

If it's fine for you, I'd take it over within this PR, which contains already a lot of desktop config migrations to system-level, which is awesome. It's only the parts that need to be installed into each users home that cause a little headache here to do right 😉.

C0D3-M4513R replied:
Ok, then go ahead. As for `graphical.target` and `multi-user.target` being reached seperately, I am also scratching my head. I think it may have something to do with having no `display-manager.service`. Another reason to abandon/depractice #4007 .
I also have not worked with user-level systemd services, so let's just cheat our way through.

MichaIng replied:
Yes, btw on that topic, related to Steam and Box86 testing, I also dealt a bit with this topic and found that there is actually no reliable method anymore to run an X server as non-root user. There is the `xserver-xorg-legacy` package which attempts to auto-detect the type of graphics driver and, if required, automatically executes X as root, which seems to have been the old default behaviour. However, it sometimes fails to detect it correctly (tested a bit up and down on VMs) and will then still fail to run X. No system group membership fixes this as it's somewhere deeper in the /dev/* directory structure. Granting render/video/tty group memberships works however for most modern graphics drivers. Last resort is to configure `xserver-xorg-legacy` a way that it always executes X as root, so does not do the test, but that breaks all security efforts behind the new attempt.

But it works fantastic to run LightDM and configure it to autologin as a certain user. It can be even configured to run any other X client application, so Steam or Amiberry or Kodi etc can all be launched through LightDM via any user without the need to deal with any sorts of graphics permissions.

I'm not able to port the way how LightDM achieves this, since it's quite complicated, but it's not necessary since LightDM is anyway mostly installed together with desktops.

I'll not implement this with v6.35 anymore, all the Box86 and Steam up and down testing costed too much time, but next release for sure.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -7165,12 +7168,13 @@ exec sudo -u $ha_user dash -c '$ha_pyenv_activation; exec pip3 install -U homeas
-			G_THREAD_START curl -sSfL https://raw.githubusercontent.com/$G_GITOWNER/DietPi/$G_GITBRANCH/.conf/desktop/lxde/pcmanfm.conf -o /root/.config/pcmanfm/LXDE/pcmanfm.conf
-			G_THREAD_START curl -sSfL https://raw.githubusercontent.com/$G_GITOWNER/DietPi/$G_GITBRANCH/.conf/desktop/lxde/pcmanfm-desktopitems.conf -o /root/.config/pcmanfm/LXDE/desktop-items-0.conf
-			G_THREAD_START curl -sSfL https://raw.githubusercontent.com/$G_GITOWNER/DietPi/$G_GITBRANCH/.conf/desktop/lxde/panel -o /root/.config/lxpanel/LXDE/panels/panel
-
```
Great to have these applied for all users. Since some of those files are part of the DEB packages themselves, we need to assure that those are marked as conffiles, so that an APT upgrade won't revert our overwrite. But otherwise I'm okay with it. Same for LXQt. I'll do a test install to verify this.

Another thing we should do is comparing our configurations with the ones from the DEB packages/defaults, since those haven't been updated for a while. Probably we can or need to do some update or we could replace a complete own config file with a few single settings changes. E.g. on LXDE I know that the panel background is an ugly image file which should be replaced with a simple grey background colour. As well the panel height is slightly increased in our config, but otherwise I'm not sure if any other differences are worth to maintain a complete own config.

C0D3-M4513R replied:
I will leave the actual config files out of this pr for now. I do not think, that I have enough experience with WM-config-files. Also that would be a whole other rabbit-hole, that would be opened.

One could try to automatically patch the config-files, after install, if it is really that important. Again, that would be another rabbit-hole, not covered by this pr. This pr is just there to apply the existing configs dynamically.

C0D3-M4513R replied:
Upon rereading this pr, I had the Idea, that this should be an issue?

MichaIng committed:
v6.35

+ DietPi-AutoStart | Do not add PCManFM desktop icon anymore. It is present as panel icon already

MichaIng committed:
v6.35

+ DietPi-Software | Desktop shortcuts are now created as follows: On first desktop install, a system-wide desktop autostart entry is created that links/copies user-level configs and desktop entries into the users home dir, once it logs in the first time. The autostart entry then disables itseld. This way only users which are actually used for desktop sessions get related configs and entries, and only a single time. Desktop shortcuts for later installed software titles are added only to users, which did already log into a desktop, hence have the autostart entry disabled. For now users, those are added by the autostart entry as well.
+ DietPi-Software | Further implement G_EXEC error handler where applicable and minor coding

MichaIng committed:
v6.35

+ DietPi-Software | DietPi-Desktop_setup: Pre-create Desktop dir in users home before attempting to link shortcuts

MichaIng committed:
v6.35

+ DietPi-Software | Satisfy shellcheck

MichaIng committed:
v6.35

+ DietPi-Software | Fix shellcheck rule ID

C0D3-M4513R commented:
You should probably mark the requested changes as complete/approved.

MichaIng commented:
"change requested" cannot be undone without approving the PR. But for that I'll now run some tests first 😉.

MichaIng committed:
v6.35

+ DietPi-Software | DietPi-Desktop_setup: Add htop to default desktop shortcuts
+ DietPi-Software | DietPi-Desktop_setup: Fix root user detection
+ DietPi-Software | DietPi-Desktop_setup: Pre-create autostart directory before writing to it
+ DietPi-Software | LXDE: Fix lxpanel default config directory and apply pcmanfm desktop item config manually within DietPi-Desktop_setup: The default desktop-items-0.conf for users cannot be overwritten by placing a file in any system directory, the content seems to be hardcoded and needs to be altered after it has been created. Since only the wallpaper and fonts shadow colour are relevant changes we apply, this is not a big issue.

MichaIng committed:
v6.35

+ DietPi-Software | LXDE: Fix applying fonts shadow colour on first login

MichaIng committed:
v6.35

+ DietPi-Software | LXQt: Fix system-wide config path

MichaIng committed:
v6.35

+ DietPi-Software | Create_Desktop_Shortcut: Fix condition

MichaIng committed:
v6.35

+ DietPi-Software | DietPi-Desktop_setup: Replace /root in bookmarks with actual users home
+ DietPi-Software | DietPi-Desktop_setup: Use desktop entry version 1.0, since MATE does not accept newest version 1.5
+ DietPi-Software | MATE: Enable audio (playback and settings panel) by installing the mate-media package
+ DietPi-Software | MATE: Skip xcompmgr install with MATE, since MATE has one integrated which blocks xcompmgr start
+ DietPi-Software | MATE: Do not add a Caja desktop icon, since by default a user home icon is present already

MichaIng committed:
v6.35

+ DietPi-Software | Desktops: Simplify xcompmgr install, which is only used/compatible with LXDE and LXQt. All other desktops have one integrated so that either the integrated one or xcompmgr fails. To be failsafe on Xfce, actively remove the autostart entry as it breaks the internal one, leading to ugly results.
+ DietPi-Software | X server: Do not install xterm by default. Each desktop has an own terminal emulator which looks and works much better. Especially when starting terminal applications from desktop entries, xterm overrides the desktops emulator, but has limited colour and fonts set, which does not look good. If there is a non-desktop X application that can make use of a terminal emulator without shipping one, we can re-add it to the related X client install.

MichaIng committed:
v6.35

+ DietPi-Software | GNUstep: Add xterm as dependency as the GPUstep "Terminal" cannot launch scripts. Remove "x-window-system-core" => "xorg" and "libc-dbg" => "libc6-dbg" dependencies which are not required to launch any of the pre-installed desktop applications.
